File: server-ts/src/services/trustEngine.ts
import { SourceType, VerificationStatus } from '../types';

export interface TrustScoreInput {
  source_type: SourceType;
  source_reliability: number; // 0 to 100
  location_confidence: number; // 0.0 to 1.0
  timestamp_age_hours: number;
  cross_source_count: number;
  ai_confidence: number; // 0.0 to 1.0
  is_duplicate: boolean;
  duplicate_score: number; // 0.0 to 1.0
}

export interface TrustEvaluationResult {
  trust_score: number; // 0 to 100
  verification_score: number; // 0 to 100
  recommended_status: VerificationStatus;
  score_breakdown: {
    source_component: number;
    location_component: number;
    timestamp_component: number;
    cross_source_component: number;
    ai_component: number;
    duplicate_penalty: number;
  };
}

export class TrustEngineService {
  /**
   * Calculates a weighted Trust Score (0-100) and Verification Status based on multi-factor signals.
   */
  public static evaluateTrust(input: TrustScoreInput): TrustEvaluationResult {
    // 1. Source Reliability (Weight 30%)
    let sourceTypeWeight = 80;
    if (input.source_type === 'GOVERNMENT') sourceTypeWeight = 98;
    else if (input.source_type === 'PUBLIC_API') sourceTypeWeight = 90;
    else if (input.source_type === 'RSS_FEED') sourceTypeWeight = 85;
    else if (input.source_type === 'SOCIAL_MEDIA') sourceTypeWeight = 65;
    else if (input.source_type === 'CITIZEN_REPORT') sourceTypeWeight = 60;

    const sourceScore = (sourceTypeWeight * 0.6) + (input.source_reliability * 0.4);
    const sourceComponent = sourceScore * 0.30;

    // 2. Location Confidence (Weight 20%)
    const locationComponent = (input.location_confidence * 100) * 0.20;

    // 3. Timestamp Freshness & Consistency (Weight 15%)
    let timestampFreshness = 100;
    if (input.timestamp_age_hours > 72) timestampFreshness = 50;
    else if (input.timestamp_age_hours > 24) timestampFreshness = 75;
    else if (input.timestamp_age_hours > 6) timestampFreshness = 90;

    const timestampComponent = timestampFreshness * 0.15;

    // 4. Cross-Source Confirmation (Weight 15%)
    // Each additional independent report boosts confidence up to 4 sources
    const crossSourceScore = Math.min(100, 40 + input.cross_source_count * 20);
    const crossSourceComponent = crossSourceScore * 0.15;

    // 5. AI Classifier Confidence (Weight 20%)
    const aiComponent = (input.ai_confidence * 100) * 0.20;

    // 6. Duplicate Penalty
    const duplicatePenalty = input.is_duplicate ? (input.duplicate_score * 30) : 0;

    // Calculate raw trust score
    let rawScore = sourceComponent + locationComponent + timestampComponent + crossSourceComponent + aiComponent - duplicatePenalty;
    const finalTrustScore = Math.max(0, Math.min(100, Math.round(rawScore * 10) / 10));

    // Verification Score computation
    const verificationScore = Math.round(finalTrustScore);

    // Determine Status
    let recommendedStatus: VerificationStatus = 'UNVERIFIED';
    if (input.is_duplicate) {
      recommendedStatus = 'DUPLICATE';
    } else if (input.source_type === 'GOVERNMENT' && finalTrustScore >= 85) {
      recommendedStatus = 'VERIFIED';
    } else if (finalTrustScore >= 82) {
      recommendedStatus = 'VERIFIED';
    } else if (finalTrustScore >= 60) {
      recommendedStatus = 'UNDER_REVIEW';
    } else if (finalTrustScore < 35) {
      recommendedStatus = 'FLAGGED';
    } else {
      recommendedStatus = 'UNVERIFIED';
    }

    return {
      trust_score: finalTrustScore,
      verification_score: verificationScore,
      recommended_status: recommendedStatus,
      score_breakdown: {
        source_component: Math.round(sourceComponent * 10) / 10,
        location_component: Math.round(locationComponent * 10) / 10,
        timestamp_component: Math.round(timestampComponent * 10) / 10,
        cross_source_component: Math.round(crossSourceComponent * 10) / 10,
        ai_component: Math.round(aiComponent * 10) / 10,
        duplicate_penalty: Math.round(duplicatePenalty * 10) / 10,
      },
    };
  }
}

File: ml/verification/trust_engine.py
"""
Trust & Verification Engine for Weather Reports.

Calculates multi-factor weighted reliability scores and assigns automated statuses:
- UNVERIFIED
- UNDER_REVIEW
- VERIFIED
- REJECTED
- FLAGGED
- DUPLICATE
"""

from typing import Dict, Any


SOURCE_RELIABILITY_WEIGHTS = {
    "GOVT_API": 0.95,
    "OPEN_DATA": 0.90,
    "RSS": 0.85,
    "PUBLIC_SOCIAL": 0.60,
    "CITIZEN": 0.50,
    "PROVIDER": 0.80
}


class TrustVerificationEngine:
    """
    Weighted Multi-Factor Trust & Verification Engine.
    """

    def __init__(self):
        self.weights = {
            "source": 0.30,
            "location": 0.20,
            "timestamp": 0.15,
            "ai_confidence": 0.15,
            "cross_source": 0.10,
            "metadata_completeness": 0.10
        }

    def calculate_trust_score(
        self,
        source_type: str,
        has_gps: bool,
        has_media: bool,
        ai_confidence: float,
        location_confidence: float,
        is_duplicate: bool = False,
        cross_source_count: int = 1
    ) -> Dict[str, Any]:
        """
        Calculates a trust score between 0.00 and 1.00.
        """
        source_score = SOURCE_RELIABILITY_WEIGHTS.get(source_type.upper(), 0.50)
        location_score = location_confidence if has_gps else (location_confidence * 0.7)
        timestamp_score = 0.90  # Valid timestamp format
        cross_source_score = min(1.0, 0.5 + (cross_source_count - 1) * 0.25)
        metadata_score = 1.0 if (has_gps and has_media) else (0.7 if (has_gps or has_media) else 0.4)

        raw_score = (
            (source_score * self.weights["source"]) +
            (location_score * self.weights["location"]) +
            (timestamp_score * self.weights["timestamp"]) +
            (ai_confidence * self.weights["ai_confidence"]) +
            (cross_source_score * self.weights["cross_source"]) +
            (metadata_score * self.weights["metadata_completeness"])
        )

        if is_duplicate:
            raw_score *= 0.85  # Penalty for duplicate status

        trust_score = round(min(1.0, max(0.0, raw_score)), 2)

        # Status determination
        if is_duplicate:
            status = "DUPLICATE"
        elif trust_score >= 0.80:
            status = "VERIFIED"
        elif trust_score >= 0.55:
            status = "UNDER_REVIEW"
        elif trust_score >= 0.35:
            status = "UNVERIFIED"
        else:
            status = "FLAGGED"

        return {
            "trust_score": trust_score,
            "verification_status": status,
            "score_breakdown": {
                "source_score": round(source_score, 2),
                "location_score": round(location_score, 2),
                "timestamp_score": round(timestamp_score, 2),
                "ai_confidence": round(ai_confidence, 2),
                "cross_source_score": round(cross_source_score, 2),
                "metadata_score": round(metadata_score, 2)
            }
        }
